File: src/java/LongestCommonPrefix.java
/**
 * 14、最长公共前缀
 * @author LJY
 */
public class LongestCommonPrefix {

    public String longestCommonPrefix(String... strs) {
        if(strs == null || strs.length == 0){
            return "";
        }
        int shortestLength = strs[0].length();
        int index = 0;
        StringBuilder commonStrBuilder = new StringBuilder(shortestLength);
        for (int i = 0; i < strs.length; i++) {
            int charsLength = strs[i].length();
            if (charsLength < shortestLength) {
                shortestLength = charsLength;
                index = i;
            }
        }
        String shortestStr = strs[index];
        for (int i = 0; i < shortestLength; i++) {
            char c = shortestStr.charAt(i);
            boolean flag = true;
            for (int j = 0; j < strs.length; j++) {
                if(j == index){
                    continue;
                }
                if(strs[j].charAt(i) != c && i == 0){
                    return "";
                }
                if ((strs[j].charAt(i) != c)) {
                    flag = false;
                }
            }
            if (flag) {
                commonStrBuilder.append(c);
            }
        }
        return commonStrBuilder.toString();
    }

    public static void main(String[] args) {
        System.out.println(new LongestCommonPrefix().longestCommonPrefix("aca","cba"));
    }

}

File: src/scala/LongestCommonPrefix.scala
object org.jy.LongestCommonPrefix {
  def longestCommonPrefix(strs: String*): String = {
    if (strs == null || strs.isEmpty) return ""
    var shortestLength = strs(0).length
    var index = 0
    val commonStrBuilder = new StringBuilder(shortestLength)
    for (i <- 0 until strs.length) {
      val charsLength = strs(i).length
      if (charsLength < shortestLength) {
        shortestLength = charsLength
        index = i
      }
    }
    val shortestStr = strs(index)
    var i = 0
    for (i <- 0 until shortestLength) {
      val c = shortestStr.charAt(i)
      var flag = true
      var j = 0
      for (j <- 0 until strs.length if j != index) {
        if (strs(j).charAt(i) != c && i == 0) return ""
        if (strs(j).charAt(i) != c) flag = false
      }
      if (flag) commonStrBuilder.append(c)
    }
    commonStrBuilder.toString
  }
}
